File: bin/heatmap_input.r
#!/usr/bin/env Rscript

library(stringr)
library(matrixStats)
library(dplyr)
library(reshape)
library(tidyr)

args = commandArgs(trailingOnly=TRUE)

if(length(args) < 1){
  stop("Usage: heatmap_input_sort20.r <tsv> <metadata> <top_n>")
}

table <- args[1]
metadata <- args[2]
top_taxa <- as.numeric(args[3])

input <- read.csv(file = table, header=TRUE, sep="\t", check.names=FALSE, stringsAsFactors=FALSE, na.strings=c("","NA"))
grouping <- read.csv(file = metadata, header=TRUE, sep="\t",check.names=FALSE, stringsAsFactors=FALSE)
input <- tidyr::separate(input, "OTU ID", c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), sep = "\\|")
rownames(grouping) <- grouping[,1]
grouping <- subset(grouping, select = c(2))
names(grouping)[1] <- "group"

filter <- function(data,grouping,top_taxa) {
  data <- aggregate(.~id,data=data,FUN=sum)
  rownames(data) <- data[,1]
  data <- subset(data, select = -c(id))
  data <- round(data, digits = 3)
  dataF <-as.data.frame(t(data))
  dataF <-cbind(dataF, grouping)
  dataF <- as.data.frame(dataF%>%group_by(group)%>%summarise_all("mean"))
  rownames(dataF) <- dataF[,1]
  dataF <- subset(dataF, select = -c(group))
  dataF<-as.data.frame(t(dataF))
  dataF$ID <- row.names(dataF)
  dataF <- melt(dataF, id.vars=c("ID"))
  dataF <- as.data.frame(dataF %>% group_by(variable) %>% slice_max(order_by = value, n = top_taxa))
  dataF <- aggregate(.~ID,data=dataF,FUN=sum)
  dataF <- subset(dataF, select = -c(variable, value))
  rownames(dataF) <- dataF[,1]
  dataC <- merge(dataF, data, by=0, all=FALSE)
  rownames(dataC) <- dataC[,1]
  dataC <- subset(dataC, select = -c(ID,Row.names))
  # dataC <- (dataC-rowMeans(dataC))/(rowSds(as.matrix(dataC)))[row(dataC)]
  # dataC <- dataC[rowSums(is.na(dataC)) == 0,]
  data_rowclust <- hclust(dist(dataC))
  data_colclust <- hclust(dist(t(dataC)))
  dataC <- dataC[data_rowclust$order, data_colclust$order]
  return(dataC)
}

species <- input[,c(6,7,8:ncol(input))]
a <- colSums(!is.na(subset(species,select =c(2))))
if (a < 10) {
  species$Species <- paste(species$Genus, species$Species)
  species$Species <- str_replace(species$Species, "NA", "sp")
  species <-species[!(species$Species=="sp NA"),]
} else {
  species <-species[!(species$Species=="NA"),]
  species$Species <- paste(species$Genus, species$Species)
  species <-species[!(species$Species=="NA NA"),]
}
species <- subset(species, select = -c(Genus))
names(species)[1] <- "id"
species <- filter(data = species,grouping = grouping, top_taxa = top_taxa)
write.table(species, "Species_taxo_heatmap.csv", sep="\t", quote=FALSE)

genus <- input[,c(6,8:ncol(input))]
names(genus)[1] <- "id"
genus <- filter(data = genus,grouping = grouping, top_taxa = top_taxa)
write.table(genus, "Genus_taxo_heatmap.csv", sep="\t", quote=FALSE)

family <- input[,c(5,8:ncol(input))]
names(family)[1] <- "id"
family <- filter(data = family,grouping = grouping, top_taxa = top_taxa)
write.table(family, "Family_taxo_heatmap.csv", sep="\t", quote=FALSE)

order <- input[,c(4,8:ncol(input))]
names(order)[1] <- "id"
order <- filter(data = order,grouping = grouping,top_taxa = top_taxa)
write.table(order, "Order_taxo_heatmap.csv", sep="\t", quote=FALSE)

class <- input[,c(3,8:ncol(input))]
names(class)[1] <- "id"
class <- filter(data = class,grouping = grouping, top_taxa = top_taxa)
write.table(class, "Class_taxo_heatmap.csv", sep="\t", quote=FALSE)

phylum <- input[,c(2,8:ncol(input))]
names(phylum)[1] <- "id"
phylum <- filter(data = phylum,grouping = grouping, top_taxa = top_taxa)
write.table(phylum, "Phylum_taxo_heatmap.csv", sep="\t", quote=FALSE)

kingdom <- input[,c(1,8:ncol(input))]
names(kingdom)[1] <- "id"
kingdom <- aggregate(.~id,data=kingdom,FUN=sum)
b <- colSums(!is.na(subset(kingdom,select =c(1))))
if (b < 2) {
  rownames(kingdom) <- kingdom[,1]
  kingdom <- subset(kingdom, select = -c(id))
  kingdom <- round(kingdom, digits = 3)
  # kingdom <- (kingdom-rowMeans(kingdom))/(rowSds(as.matrix(kingdom)))[row(kingdom)]
  # kingdom <- kingdom[rowSums(is.na(kingdom)) == 0,]
  write.table(kingdom, "Kingdom_taxo_heatmap.csv", sep="\t", quote=FALSE)
} else {
  kingdom <- filter(data = kingdom,grouping = grouping, top_taxa = top_taxa)
  write.table(kingdom, "Kingdom_taxo_heatmap.csv", sep="\t", quote=FALSE)
}

File: bin/heatmap_input.py
#!/usr/bin/env python

import argparse
import pandas as pd
import numpy as np
import re
import seaborn as sns

def extract_taxa(taxa_string, index):
    unidentified_taxa_patterns = [
        r"^$",
        r"metagenome",
        r"uncultured",
        r"unidentified"
    ]
    levels = taxa_string.split(";")
    levels = [re.sub(r"^\w?__", "", l.strip()) for l in levels]
    if levels[index]:
        result = levels[index]
    else:
        result = "sp."
    while any(re.search(p, levels[index], re.IGNORECASE) for p in unidentified_taxa_patterns) and index>0:
        index -= 1
        if levels[index]:
            result = levels[index] + " " + result
    return result

def make_heatmap_input(reltax_tables, metadata, top_n):
    levels = {
        "level-1.csv": ("Kingdom", 0),
        "level-2.csv": ("Phylum", 1),
        "level-3.csv": ("Class", 2),
        "level-4.csv": ("Order", 3),
        "level-5.csv": ("Family", 4),
        "level-6.csv": ("Genus", 5),
        "level-7.csv": ("Species", 6)
    }

    # Read metadata file
    meta = pd.read_csv(metadata, sep="\t")
    for table in reltax_tables:
        try:
            level, index = levels[table]
        except:
            print("{} not recoginzed as one of the relative abundance table!".format(table))
            continue
        data = pd.read_csv(table)
        # Conver to fractions
        num_cols = data.select_dtypes(include=np.number).columns
        data[num_cols] = data[num_cols].div(data[num_cols].sum())
 
        # Merge with group labels
        data_merge = pd.melt(data, id_vars="#OTU ID", var_name="sampleid", value_name="abundance")
        data_merge = data_merge.merge(meta)

        # Get top n most abundant taxa from each group
        top_taxa = data_merge.groupby(["#OTU ID", "group"])["abundance"].mean().groupby("group").nlargest(top_n)
        top_taxa = top_taxa.droplevel(0).reset_index()["#OTU ID"].unique()

        # Subset abundance data for only top n taxa
        data = data.loc[data["#OTU ID"].isin(top_taxa), ]

        # Extract taxonomy for just this level + uppper levels if necessary
        data["#OTU ID"] = data["#OTU ID"].apply(extract_taxa, index=index)

        # Only keep samples that are in the metadata
        data = data.set_index("#OTU ID")
        data = data.loc[:, data.columns.isin(meta["sampleid"])]

        # Clustering (only if at least 5 taxa)
        if len(data) >= 5:
            # Get the lowest count for pseudocount
            if (data>0).all().all():
                pseudocount = 0
            else:
                pseudocount = data.apply(lambda x:x[x>0].min()).min() / 2
            # Log transform (data+pseudocount)
            transformed = np.log10(data+pseudocount)
            fig = sns.clustermap(transformed)
            data = data.iloc[fig.dendrogram_row.reordered_ind, fig.dendrogram_col.reordered_ind]
            data.to_csv("{}_taxo_heatmap.csv".format(level), sep="\t")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="""Parse and filter relative abundance data for making heatmap""")
    parser.add_argument(dest="reltax_tables", nargs="+", type=str, help="Relative abundance tables")
    parser.add_argument("-m", "--metadata", dest="metadata", required=True, help="Metadata file")
    parser.add_argument("-t", "--top_n", dest="top_n", type=int, default=20, help="")
    args = parser.parse_args()
    make_heatmap_input(args.reltax_tables, args.metadata, args.top_n)